chore(deps): upgrades jest to 28

dimitropoulos opened the pull request:
In lieu of the normal description, I have written a bunch of comments on specific topics that came up during this work.  The one and only goal of this PR is to upgrade jest with the smallest footprint possible.  If you have any ideas on how to further reduce the footprint, lemme know!

dimitropoulos committed:
update jest and ts-jest

dimitropoulos committed:
move `mocked` imports (now deprecated in `ts-jest`) to `jest-mock`

dimitropoulos committed:
adds optional chaining to insomnia-app babel loader

I really have little or no idea why this is failing without this (and what effect simply updating jest had on this) but... so it goes.

dimitropoulos committed:
wip

dimitropoulos committed:
wip

dimitropoulos committed:
wip

dimitropoulos committed:
something is going on with RTL

dimitropoulos committed:
and then there were 7.  back to buffer polyfill needs

dimitropoulos committed:
makes localstorage tests use fake jest timers so they pass

dimitropoulos committed:
all but 1

dimitropoulos committed:
removes old filesystem migrations

dimitropoulos committed:
fixes buffer for crypt

dimitropoulos committed:
lockfiles

dimitropoulos committed:
self-review

dimitropoulos committed:
more self review

dimitropoulos commented on packages/insomnia-config/src/validate.test.ts:
```diff
@@ -1,3 +1,5 @@
+import { describe, expect, it } from '@jest/globals';
```
after this PR, we can go ahead and add this to every test file in the project, which will allow us to no longer depend on the globals and also will mean that TypeScript will correctly tell us we cannot use jest stuff from our source code (by accident, for example).

marckong replied:
ok I see. I actually like the clarity. Thank you for doing this work.

jackkav replied:
This would make a good low risk introductory refactoring PR to get familiar with the codebase as a whole.  :wink: @marckong

dimitropoulos replied:
for anyone interested to learn more about `@jest/globals` https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44365 is a decent primer.  the TLDR; is that it allows us to use jest without globals (which, for us, means that we can't then accidentally call jest globals from our non-test source code).

dimitropoulos commented on jest-preset.js:
```diff
@@ -3,7 +3,7 @@ module.exports = {
   globals: {
     'ts-jest': {
-      isolatedModules: true,
+      isolatedModules: false,
```
it's time for this default to flip.  it means our tests will be typechecked by default (which, then means we'll have to turn it off in those that have errors rather than turn it on in those that don't, and new packages will have it by default).

dimitropoulos commented on packages/insomnia-components/src/tooltip.test.tsx:
```diff
@@ -1,10 +1,9 @@
 import { Tooltip } from './tooltip';
 
-const pause = (time: number) => new Promise(resolve => setTimeout(resolve, time));
-
```
this test was actually pausing in real wall-clock time before.  now.. it doesn't.

dimitropoulos commented on packages/insomnia-inso/tsconfig.build.json:
```diff
@@ -13,7 +13,6 @@
     "node_modules",
     "dist",
     "src/jest",
-    ".babelrc.js",
```
this file was deleted a while back and it looks like we just missed deleting it here.

dimitropoulos commented on packages/insomnia/.babelrc.js:
```diff
@@ -1,15 +0,0 @@
-/** @type { import('@babel/core').TransformOptions } */
```
this is unused insofar as I can tell.  I thought maybe jest uses it, but I tested that and it appears it does not.

dimitropoulos commented on packages/insomnia/jest.config.js:
```diff
@@ -1,29 +1,20 @@
 /** @type { import('@jest/types').Config.InitialOptions } */
 module.exports = {
-  // preset: '../../jest-preset.js', // DOES NOT WORK
```
I figured out why this doesn't work see https://github.com/facebook/jest/pull/12817 for what the solution was

dimitropoulos commented on packages/insomnia/jest.config.js:
```diff
@@ -1,29 +1,20 @@
-    'styled-components': '<rootDir>/../node_modules/styled-components',
+    '\\.(css|less|png)$': '<rootDir>/src/__mocks__/dummy.ts',
+    'styled-components': '<rootDir>/node_modules/styled-components',
+    'jsonpath-plus': '<rootDir>/node_modules/jsonpath-plus/dist/index-node-cjs.cjs',
```
it may not seem like much, but this line of code took me about 5 hours, maybe more, to arrive at.  there's some funky business for jsonpath-plus that we have to work around and it was quite a bear.

jackkav replied:
Related reference: https://nodejs.org/api/packages.html#conditional-exports
Scoped out of this PR are changes to testEnvironment: jsdom
Which resolves `"browser"` modules by default, rather than the commonjs `"node"` modules which would be resolved in electron context currently. Also related to electron api `"contextIsolation"`

marckong commented on packages/insomnia/src/common/__tests__/misc.test.ts:
```diff
@@ -142,22 +142,27 @@ describe('filterHeaders()', () => {
+  it('debounces correctly', async () => {
+    jest.useFakeTimers();
+    const resultsList: Record<string, string[]>[] = [];
+    console.log('starting'.repeat(100));
```
I am wondering if this console.log is intended?

dimitropoulos replied:
I removed it in a subsequent self-review.  you were fast! I hadn't pulled this out of draft yet because I was doing a self review still. :)

dimitropoulos commented on packages/insomnia/src/account/__tests__/crypt.test.ts:
```diff
@@ -33,9 +35,10 @@ describe('crypt', () => {
+      const resultEncrypted = crypt.encryptAESBuffer(key, source);
       const resultDecrypted = crypt.decryptAESToBuffer(key, resultEncrypted);
-      expect(resultDecrypted).toEqual(Buffer.from('Hello World!', 'utf8'));
+      expect(resultDecrypted.toString()).toEqual(source.toString());
```
it's non-ideal to have to `toString` these before comparison, but if we don't the match fails because one buffer comes from the npm and one comes from node.  and if we try to flip _that_ around we get stuck because we can't polyfill what's in the `hkdf` npm (which doesn't use the `buffer` npm, and uses the global).

marckong commented on packages/insomnia/src/models/__tests__/experiment.test.ts:
```diff
@@ -0,0 +1,14 @@
+  writeFileSync('/tmp/experiment', 'some data');
+};
+
+describe('experiment', () => {
```
I wonder if this experiment was only for the experimenting some code

dimitropoulos replied:
see https://github.com/Kong/insomnia/pull/4783#discussion_r872611127 but good to know you've got a keen eye!

dimitropoulos commented on packages/insomnia/src/common/__tests__/local-storage.test.ts:
```diff
@@ -103,21 +94,16 @@ describe('LocalStorage()', () => {
-    expect(setTimeout).toHaveBeenCalledTimes(3);
-    expect(setTimeout).nthCalledWith(1, expect.anything(), 100);
-    expect(setTimeout).nthCalledWith(2, expect.anything(), 100);
-    expect(setTimeout).nthCalledWith(3, expect.anything(), 100);
```
for these kinds of things it was tricky and everyone on the team agreed that asserting on the result is plenty verification.  we don't need to mock and assert on the internals (e.g. setInterval).  I still wish I understood why it didn't work, but so it goes.

dimitropoulos commented on packages/insomnia/src/common/__tests__/validate-insomnia-config.test.ts:
```diff
@@ -1,4 +1,4 @@
-import { mocked } from 'ts-jest/utils';
+import { mocked } from 'jest-mock';
```
this happened in ts-jest 27.1.0.  see https://github.com/kulshekhar/ts-jest/blob/main/CHANGELOG.md#2710-2021-11-30

dimitropoulos commented on packages/insomnia/src/common/database.ts:
```diff
@@ -348,7 +342,9 @@ export const database = {
         ),
       );
-      collection.persistence.setAutocompactionInterval(DB_PERSIST_INTERVAL);
+      if (!config.inMemoryOnly) {
```
this should be an effective noop in runtime and has the benefit of helping out the tests

dimitropoulos committed:
yet more self review

marckong reviewed (approved):
LGTM. We've paired on the Zoom with the team, so I've seen this works like a charm! I am just wondering if console log and experiment.test were intended to be included.

dimitropoulos commented:
@marckong see https://github.com/Kong/insomnia/pull/4783#discussion_r872611127.  I was doing a self review (actually I went through the PR 3 separate times top-to-bottom and found things each time.  The signal that it's ready to review would be pulling it out of draft and assigning reviewers, which I will do immediately after typing this message. :)

marckong commented:
Yeah I didn't realize this was the draft lol I automatically jumped onto it because it reminded me of the zoom call excitement! Amazing work!

dimitropoulos committed:
removes now-unneeded jasmine jest setup

e.g. https://github.com/Kong/insomnia/commit/5fa8f0069d10db0a847abfaececc8c9ed9197275#diff-7694d151b00a40400e2d04f58556e8f8991d01ed34b747d912c8158fa381b509L96

no longer used

dimitropoulos committed:
satisfy linter

filfreire reviewed (approved):
Tested it locally for a while, LGTM 👍

jackkav reviewed (approved):
- 25k package lock, very nice.
isolatedModules might be something we come back to later, hopefully no sharp edges. :crossed_fingers:
